include/internal/quic_record_tx.h

```diff
@@ -0,0 +1,299 @@
+    const unsigned char    *buf;
+    size_t                  buf_len;
+} OSSL_QTX_IOVEC;
+
+typedef struct ossl_qtx_pkt_st {
+    /* Logical packet header to be serialized. */
+    QUIC_PKT_HDR               *hdr;
+
+    /*
+     * iovecs expressing the logical packet payload buffer. Zero-length entries
+     * are permitted.
+     */
+    const OSSL_QTX_IOVEC       *iovec;
+    size_t                      num_iovec;
+
+    /* Destination address. Will be passed through to the BIO if non-NULL. */
+    const BIO_ADDR             *peer;
+
+    /*
+     * Local address (optional). Specify as non-NULL only if TX BIO
+     * has local address support enabled.
+     */
+    const BIO_ADDR             *local;
+
+    /*
+     * Logical PN. Used for encryption. This will automatically be encoded to
+     * hdr->pn, which need not be initialized.
+     */
+    QUIC_PN                     pn;
+
+    /* Packet flags. Zero or more OSSL_QTX_PKT_FLAG_* values. */
+    uint32_t                    flags;
+} OSSL_QTX_PKT;
+
+/*
+ * More packets will be written which should be coalesced into a single
+ * datagram; do not send this packet yet. To use this, set this flag for all
+ * packets but the final packet in a datagram, then send the final packet
+ * without this flag set.
+ *
+ * This flag is not a guarantee and the QTX may transmit immediately anyway if
+ * it is not possible to fit any more packets in the current datagram.
+ *
+ * If the caller change its mind and needs to cause a packet queued with
+ * COALESCE after having passed it to this function but without writing another
+ * packet, it should call ossl_qtx_flush_pkt().
+ */
+#define OSSL_QTX_PKT_FLAG_COALESCE       (1U << 0)
+
+/*
+ * Writes a packet.
+ *
+ * *pkt need be valid only for the duration of the call to this function.
+ *
+ * pkt->hdr->data and pkt->hdr->len are unused. The payload buffer is specified
+ * via an array of OSSL_QTX_IOVEC structures. The API is designed to support
+ * single-copy transmission; data is copied from the iovecs as it is encrypted
+ * into an internal staging buffer for transmission.
+ *
+ * The function may modify and clobber pkt->hdr->data, pkt->hdr->len,
+ * pkt->hdr->key_phase and pkt->hdr->pn for its own internal use. No other
+ * fields of pkt or pkt->hdr will be modified.
+ *
+ * It is the callers responsibility to determine how long the PN field in the
+ * encoded packet should be by setting pkt->hdr->pn_len. This function takes
+ * care of the PN encoding. Set pkt->pn to the desired PN.
+ *
+ * Note that 1-RTT packets do not have a DCID Length field, therefore the DCID
+ * length must be understood contextually. This function assumes the caller
+ * knows what it is doing and will serialize a DCID of whatever length is given.
+ * It is the caller's responsibility to ensure it uses a consistent DCID length
+ * for communication with any given set of remote peers.
+ *
+ * The packet is queued regardless of whether it is able to be sent immediately.
```

Comment:
Perhaps add some comment why? I.E., to support use of the sendmmsg().

Reply:
Will do.